overlord/devicestate/devicemgr.go

```diff
@@ -82,6 +82,12 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	runner.AddHandler("mark-seeded", m.doMarkSeeded, nil)
 	// this *must* always run last and finalizes a remodel
 	runner.AddHandler("set-model", m.doSetModel, nil)
+	// XXX: gadget update is tricky, and currently does not support undo.
+	// The system is rebooted during update, if it boots up to the point
+	// where snapd runs we deem the new assets (be it bootloader or
+	// firmware) functional. However, once gadget snap is reverted, the
+	// on-disk state may not be in sync with the description in gadget.yaml.
+	runner.AddHandler("update-gadget", m.doUpdateGadget, nil)
```

Comment:
what's the latest thinking on undo here, @mvo5 as well ?

Reply:
From last chat we had, the problem is that gadget data does not have similar immutability guarantees as snaps do. The whole gadget update is kind of out of band. Once we reboot, the boot process may have modified the data/configuration/environment in a way that makes it incompatible with previous revision of the snap.

The idea is that once a reboot is successful and we reach a stage when snapd runs, the boot assets are deemed correct (or functional at least). Even if there is a problem in later tasks, boot assets should stay. The fix, if any, would come in the next revision of a snap, which would traverse the whole update path again.

Reply:
I don't think keeping a XXX here is useful, but we need to have a comment that illustrates what we decided because we will need to document that for the users of the feature.  Basically we don't undo fully succesfull gadget assets updates, but assume that new assets will be backward compatible enough do work even if the rest of gadget or kernel themselves get reverted? is that correct (cc @mvo). There is also the question of what `snap revert`on gadget should do but that is more related to once the task is actually used.

Reply:
I've updated the comment.